### DISCUSSION TITLE
Primary constructors in their current state are horrible

### DISCUSSION THREAD

**Discussion by jonashendrickx** [dotnet/csharplang#7937] [Language Ideas]
We should be able to make parameters of primary constructors `protected`, `private` or `readonly`. And possibly also enforce a consistent syntax, i.e. `_camelCase` for private fields, `PascalCase` for protected/public fields.

`public class TellTale(private readonly string Id, readonly string Title);`

`public class ServiceA(private readonly IDependencyA, private readonly IDependencyB);`

This would also address the inconsistency with classes where parameters of primary constructors are by default private and mutable, while with records they are public but immutable.

**Comment by yaakov-h** [comment]
> And possibly also enforce a consistent syntax, i.e. `_camelCase` for private fields, `PascalCase` for protected/public fields.

The language does not guarantee that there will be any fields.

**Comment by chrisoverzero** [comment]
> We should be able to make parameters of primary constructors `protected`, `private` or `readonly`.

What would it mean for a _parameter_ to be `protected`?

> And possibly also enforce a consistent syntax, i.e. `_camelCase` for private fields, `PascalCase` for protected/public fields.

You can assign the parameters from primary constructors to fields with whatever naming scheme you want, as enforceable by existing tools. The parameters themselves are not fields, they are parameters.

> This would also address the inconsistency with classes […]

Non-record classes are different from record classes – record classes do more, by design. If they behaved the same, what would the point be of having both of them?

**Comment by tacosontitan** [comment]
I think you're confusing primary constructor *parameters* for *fields*. The parameters are not guaranteed to become fields, but if you want fields, then you can always explicitly map them like you would for any other constructor:

```cs
public class Sample(int foo)
{
    private readonly int _foo = foo;
    ...
}
```

This will ensure that you've mapped `foo` to a field and satisfies your requested accessibility constraints.

`foo` is a *parameter* and should be named and utilized as such. Constructors (primary or not) are just a special method for creating a new instance of your object, so even typical constructors have parameters that you can utilize directly and discard or map to fields:

```cs
public class Sample
{
    private readonly int _foo;
    private readonly IBarService _barService;
    public Sample(int foo, IServiceProvider serviceProvider)
    {
        _foo = foo;
        _barService = serviceProvider.GetRequiredService<IBarService>();
    }
}
```

In this example, `serviceProvider` does not become a field, nor would it when used the same way with a primary constructor:

```cs
public class Sample(int foo, IServiceProvider serviceProvider)
{
    private readonly int _foo = foo;
    private readonly IBarService _barService = serviceProvider.GetRequiredService<IBarService>();
}
```

As [this SharpLab snippet](https://sharplab.io/#v2:D4AQTAjAsAULIGYAEBLAdgFwKYCcBmAhgMZZICSAsgRgBYDKuAbiiUgN6qZIUCuANhhQAHPgE8AFOgxICAGk7SARgEoA3EgC+8ZFNyFWZAEIEcDHM1YcQAFiQAFHFIA8AFQB84l5yE8MazdoKesSkZHQEALYiWA4A9swAJrjsSF4A4lgYAEpYAI48KDhYCWYWWK4e/lpwMAD09UhOAM48EREmom6w9bVIOUJFTViYTTJITZHRSERF1MVIAO4otGMDKO04otOxaE0YODxEGLE4AHTdDU61LW0dXTCISOBI4VF8MY4bErBIvwpIeFisVkPz+lGo9CYLFI7VopWhIJgf3Ir2icUSySG5mh6JQSRwylgbFBvzWjDmSFmCR2YnIxlMUNYAH1FCZ4awALzjRkfeJ43CnDLZPIFIolHlOIxsnmVVQkpDyx42eyOTDiQlIv7EzXI37knBIWK+HzSLmwyHYkinXgCYRicSA4FIBBqeXIlnSy1YU4OKTiI0YE2unUBGDVC69ZqtL73Hp9LADLBDEZjCZvUgzLBzBKLZY0Mb7AgJZYoHYEPjbXb7Q7HM4RxrXaN3QLPVHvFw4Islst8InyskUqk0rZSJBMx1ykMD7CUrPUtC0qUMr1j1nLsqT5HT0hDhdbcFwnlj83srCTxXINtYDtdwQ98Ruv6jx2I3VgqiHlcnnmvt8oybvLi+Lcl6QG4BqyLan+Y6OkgXITo+vxMt+K5mhCp6bn+HrrtCcEgWUYFnEKOT5IUxSnpK9KnrK8rVMiF5PLYvpqhBWqIUg+qGsavh4ch6E8ta/CCCIEjjkC8guphb7YaePqqhg/rcX4UmhhoQA==) demonstrates:

![image](https://github.com/dotnet/csharplang/assets/65432314/f6c07adf-1e8a-437b-b42e-c9a0fc31f922)


Primary constructors help us simplify our constructors and model definitions by giving us a way to extend the range available to their parameters.

**Reply by jonashendrickx** [reply]
![image](https://github.com/dotnet/csharplang/assets/1045327/3c12f1b0-c257-4eb4-a6fd-7cea65c96a71)

![image](https://github.com/dotnet/csharplang/assets/1045327/2b25fc51-4e45-4b0d-8f42-0d3451cb86a8)

I tested your suggestion. But with value types it can still to turn into a nightmare.

With reference types, you can still modify the parameter, and your readonly field will still update.

![image](https://github.com/dotnet/csharplang/assets/1045327/a64d8693-1ab7-4e92-9e67-ce71f2b5a661)

![image](https://github.com/dotnet/csharplang/assets/1045327/4fb4898a-3266-404e-a87f-f031bb9739c3)

I like the implementation how it was done in Kotlin. But in .NET it can turn into an absolute nightmare.

**Reply by chrisoverzero** [reply]
> I tested your suggestion.

Don't mix use of a parameter and a field which has had that parameter assigned to it. The compiler caught this for you – see the green smudge under the assignment to `_wheels`?

> Parameter 'ushort wheels' is captured into the state of the enclosing type and its value is also used to initialize a field, property, or event.

That's the compiler's saying "probably don't do that". Pick one or the other.

**Comment by colejohnson66** [comment]
Perhaps a source generator+analyzer could be created. The generator would take all (or a subset of) parameters and store them into backing fields following the desired naming convention (`paramA` -&gt; `_paramA` or `m_paramA`). The analyzer would warn (or error) on usages of the parameters (outside its own, obviously). So, the generator would do the following:

```cs
// your code
[PrimaryConstructorToFields(NamingConvention = PrimaryConstructorNamingConventions.UnderscorePrefix)]
public partial class MyClass(int value)
{
    public int GetValue =>
        value; // WARNING: Do not use primary constructor parameters; Use `_value` instead.
}

// generated
partial class MyClass
{
    private int _value = value;
}
```

It could even get fancy and skip making fields for ones you've already made, or even make the ones it adds `readonly`:

```cs
// your code
[PrimaryConstructorToFields(NamingConvention = PrimaryConstructorNamingConventions.MUnderscorePrefix)]
[PrimaryConstructorReadonlyFields(nameof(value))]
public partial class MyClass(string name, int value)
{
    private string _name = name ?? throw new ArgumentNullException(nameof(name));

    public string Name =>
        _name;
    public int Value =>
        m_value;
}

// generated
partial class MyClass
{
    private readonly int m_value = value;
}
```
